File: example/src/test/java/com/example/service/ExampleServiceTest.java
package com.example.service;

import com.example.base.TestBase;
import com.example.model.User;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.Assert;

import java.util.List;

public class ExampleServiceTest extends TestBase {
    private final ExampleService exampleService;

    @Autowired
    ExampleServiceTest(ExampleService exampleService) {
        this.exampleService = exampleService;
    }

    @Test
    public void findUsersTest() {
        List<User> userList = exampleService.findUsers();
        Assert.notEmpty(userList);
    }
}

File: example-kotlin/src/test/kotlin/com/example/service/ExampleServiceTest.kt
package com.example.service

import com.example.base.TestBase
import com.example.model.UserDetail
import com.example.model.UserMain
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Order

class ExampleServiceTest(
    var exampleServiceImpl: ExampleService
) : TestBase() {

    companion object {
        private const val USERNAME = "Gary"
    }

    @BeforeAll
    fun `insert new user`() {
        val main = UserMain(name = USERNAME).run {
            exampleServiceImpl.saveOrUpdateUser(this)
        }

        UserDetail().run {
            this.name = main.name
            this.gender = "M"
            this
        }.run {
            exampleServiceImpl.saveOrUpdateInfo(this)
        }
    }

    @AfterAll
    fun `delete user`() {
        val main = exampleServiceImpl.findUserByName(USERNAME)
        exampleServiceImpl.deleteUser(main!!)
        val detail = exampleServiceImpl.findUserInfoByName(main.name)
        exampleServiceImpl.deleteInfo(detail!!)
    }

    @Test
    @Order(1)
    fun `find user by username`() {
        val main = exampleServiceImpl.findUserByName(USERNAME)
        Assertions.assertNotNull(main)
    }

    @Test
    @Order(2)
    fun `update user`() {
        val main = exampleServiceImpl.findUserByName(USERNAME) ?: throw Exception()

        val gender = "F"
        exampleServiceImpl.findUserInfoByName(main.name)?.let {
            it.gender = gender
            exampleServiceImpl.saveOrUpdateInfo(it)
        }

        val detail = exampleServiceImpl.findUserInfoByName(main.name)
        Assertions.assertEquals(gender, detail?.gender)
    }

    @Test
    @Order(3)
    fun `find user by jdbc`() {
        val main = exampleServiceImpl.findUserByJdbc(USERNAME)
        Assertions.assertNotNull(main)
    }
}
